### PR TITLE
[core] update Child Update Request aggregation feature

### PR DESCRIPTION
**Pull request by librasungirl**
To introduce additional time delay for aggregating Child Update Request Messages as discussed in #2858, this commit replaces the task with a timer which has a fixed aggregation delay (1s).

### CODE REVIEW

File: src/core/thread/mle.hpp [hunk 1]
```diff
@@ -1582,6 +1582,13 @@ class Mle : public InstanceLocator
         kParentRequestTypeRoutersAndReeds, ///< Parent Request to all routers and REEDs.
     };
 
+    enum ChildUpdateRequestPendingStatus
+    {
+        kChildUpdateRequestPendingNo,
+        kChildUpdateRequestPendingInProgress,
+        kChildUpdateRequestPendingInit,
```
Comment (jwhui) [inline src/core/thread/mle.hpp]: nit: what about
```c++
enum ChildUpdateRequestState
{
    kChildUpdateRequestNone,
    kChildUpdateRequestPending,
    kChildUpdateRequestActive,
};
```

I also tend to prefer to list the states in the order that they typically occur.

File: src/core/thread/mle.cpp [hunk 2]
```diff
@@ -2037,33 +2043,32 @@ otError Mle::SendDataRequest(const Ip6::Address &aDestination,
...
-}
 
-void Mle::HandleChildUpdateRequestTimer(void)
-{
-    if (mParent.IsStateRestoring() || IsRxOnWhenIdle())
+    if (mChildUpdateRequestPendingStatus == kChildUpdateRequestPendingInProgress)
+    {
+        SendChildUpdateRequest();
+    }
+    else if (mParent.IsStateRestoring() || IsRxOnWhenIdle())
```
Comment (jwhui) [inline src/core/thread/mle.cpp]: nit: combine with if-statement above?

File: src/core/thread/mle_constants.hpp [hunk 3]
```diff
@@ -57,22 +57,23 @@ enum
...
+    kParentRequestRouterTimeout     = 750,   ///< Router Parent Request timeout
+    kParentRequestReedTimeout       = 1250,  ///< Router and REEDs Parent Request timeout
+    kAttachStartJitter              = 50,    ///< Maximum jitter time added to start of attach.
+    kAnnounceProcessTimeout         = 250,   ///< Timeout after receiving Announcement before channel/pan-id change
+    kAnnounceTimeout                = 1400,  ///< Total timeout used for sending Announcement messages
+    kMinAnnounceDelay               = 80,    ///< Minimum delay between Announcement messages
+    kParentResponseMaxDelayRouters  = 500,   ///< Maximum delay for response for Parent Request sent to routers only
+    kParentResponseMaxDelayAll      = 1000,  ///< Maximum delay for response for Parent Request sent to all devices
+    kUnicastRetransmissionDelay     = 1000,  ///< Base delay before retransmitting an MLE unicast.
+    kChildUpdateRequestPendingDelay = 1000,  ///< Delay (in ms) for aggregating Child Update Request.
```
Comment (jwhui) [inline src/core/thread/mle_constants.hpp]: Do we need to wait 1 second?  Can this be smaller?

Comment (librasungirl) [reply src/core/thread/mle_constants.hpp]: shorted to 100ms in new patch. is it ok?

File: src/core/thread/mle.hpp (hunk 1 above)
Comment (librasungirl) [reply src/core/thread/mle.hpp]: updated.

File: src/core/thread/mle.cpp (hunk 2 above)
Comment (librasungirl) [reply src/core/thread/mle.cpp]: updated.

File: src/core/thread/mle.cpp [hunk 4]
```diff
@@ -1484,6 +1485,11 @@ void Mle::HandleStateChanged(otChangedFlags aFlags)
 #endif
     }
 
+    if (mChildUpdateRequestState == kChildUpdateRequestPending)
+    {
+        mChildUpdateRequestTimer.Start(kChildUpdateRequestPendingDelay);
```
Comment (abtink) [inline src/core/thread/mle.cpp]: How about if we limit the restart of the timer only to `OT_CHANGED_<>` events  that require a child update to be sent?
My concern is that unrelated notification events may restart the timer and delay the tx of child update.

Thoughts?

File: src/core/thread/mle.cpp [hunk 5]
```diff
@@ -2089,7 +2093,8 @@ otError Mle::SendChildUpdateRequest(void)
 
     if (mChildUpdateAttempts >= kMaxChildKeepAliveAttempts)
     {
-        mChildUpdateAttempts = 0;
+        mChildUpdateAttempts     = 0;
+        mChildUpdateRequestState = kChildUpdateRequestNone;
         BecomeDetached();
         ExitNow();
     }
```
Comment (abtink) [inline src/core/thread/mle.cpp]: How about setting `mChildUpdateRequest = kChildUpdateRequestActive` after this `if`?

I think this change can fix an bug we currently have in the code where on SED, "Child Update" is attempted once and after the timeout, a "Data Request" is sent (please see current [`HandleChildUpdateRequestTimer`](https://github.com/openthread/openthread/blob/f0654381117d6f515e5808a1750761fe2719c5a4/src/core/thread/mle.cpp#L2064-L2081)).

If we change state to `Active`, when other modules call `Mle::SendChildUpdateRequest()`, if no response is received after timeout we will know to send a "Child Update".

File: src/core/thread/mle.cpp (hunk 4 above)
Comment (librasungirl) [reply src/core/thread/mle.cpp]: make sense.

File: src/core/thread/mle.cpp (hunk 5 above)
Comment (librasungirl) [reply src/core/thread/mle.cpp]: Agree with the proposal. will update as your suggestion.

File: src/core/thread/mle.cpp [hunk 6]
```diff
@@ -1484,6 +1485,12 @@ void Mle::HandleStateChanged(otChangedFlags aFlags)
 #endif
     }
 
+    if (mChildUpdateRequestState == kChildUpdateRequestPending)
+    {
+        mChildUpdateAttempts = 0;
+        mChildUpdateRequestTimer.Start(kChildUpdateRequestPendingDelay);
```
Comment (abtink) [inline src/core/thread/mle.cpp]: How about moving the above two lines to after where we set the `mChildUpdateRequestState = kChildUpdateRequestPending;` in two cases above in  this method?

This way, we ensure that the timer is only restarted when the state does change. The reason is if we receive a `HandeStateChanged` with another flag (that does not require a "Child Update") while we are in Pending state, I don't think we need to restart the timer.

Thoughts?

Comment (librasungirl) [reply src/core/thread/mle.cpp]: oh. yes. it make senses.

**Pull request merged**